File: routes/categories.js
const express = require('express');
const router = express.Router();
const Category = require('../models/Category');
const Todo = require('../models/Todo');

// ─── GET /api/categories ──────────────────────────────────────────────────────
router.get('/', async (req, res) => {
  try {
    const categories = await Category.find().sort({ name: 1 });
    res.json({ success: true, data: categories });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
});

// ─── POST /api/categories ─────────────────────────────────────────────────────
router.post('/', async (req, res) => {
  try {
    const { name, color, icon } = req.body;
    if (!name) return res.status(400).json({ success: false, message: 'Category name is required' });

    const category = new Category({ name: name.trim(), color, icon });
    const saved = await category.save();
    res.status(201).json({ success: true, data: saved });
  } catch (err) {
    if (err.code === 11000) {
      return res.status(400).json({ success: false, message: 'Category name already exists' });
    }
    res.status(400).json({ success: false, message: err.message });
  }
});

// ─── PUT /api/categories/:id ──────────────────────────────────────────────────
router.put('/:id', async (req, res) => {
  try {
    const category = await Category.findByIdAndUpdate(req.params.id, req.body, {
      new: true,
      runValidators: true,
    });
    if (!category) return res.status(404).json({ success: false, message: 'Category not found' });
    res.json({ success: true, data: category });
  } catch (err) {
    res.status(400).json({ success: false, message: err.message });
  }
});

// ─── DELETE /api/categories/:id ───────────────────────────────────────────────
router.delete('/:id', async (req, res) => {
  try {
    const category = await Category.findByIdAndDelete(req.params.id);
    if (!category) return res.status(404).json({ success: false, message: 'Category not found' });

    // Unlink todos from deleted category
    await Todo.updateMany({ category: req.params.id }, { category: null });

    res.json({ success: true, message: 'Category deleted and todos unlinked' });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
});

module.exports = router;

File: routers/categories.py
"""
routers/categories.py
─────────────────────
CRUD endpoints for Task Categories.

Endpoints:
    GET    /api/categories        → list all
    POST   /api/categories        → create
    PUT    /api/categories/{id}   → update
    DELETE /api/categories/{id}   → delete (also unlinks todos)
"""

from beanie import PydanticObjectId
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from models.category import Category
from models.todo import Todo

router = APIRouter()


# ── Schemas ───────────────────────────────────────────────────────────────────

class CategoryCreate(BaseModel):
    name:  str
    color: str  = "#6366f1"
    icon:  str  = "📁"


class CategoryUpdate(BaseModel):
    name:  Optional[str] = None
    color: Optional[str] = None
    icon:  Optional[str] = None


# ── Helpers ───────────────────────────────────────────────────────────────────

def _serialize(cat: Category) -> dict:
    return {
        "id":    str(cat.id),
        "name":  cat.name,
        "color": cat.color,
        "icon":  cat.icon,
    }


async def _get_or_404(cat_id: str) -> Category:
    try:
        oid = PydanticObjectId(cat_id)
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid category id")
    cat = await Category.get(oid)
    if not cat:
        raise HTTPException(status_code=404, detail="Category not found")
    return cat


# ── GET /api/categories ───────────────────────────────────────────────────────

@router.get("/")
async def list_categories():
    cats = await Category.find_all().sort(+Category.name).to_list()
    return {"success": True, "data": [_serialize(c) for c in cats]}


# ── POST /api/categories ──────────────────────────────────────────────────────

@router.post("/", status_code=201)
async def create_category(body: CategoryCreate):
    if not body.name.strip():
        raise HTTPException(status_code=400, detail="Category name is required")

    existing = await Category.find_one(Category.name == body.name.strip())
    if existing:
        raise HTTPException(status_code=400, detail="Category name already exists")

    cat = Category(name=body.name.strip(), color=body.color, icon=body.icon)
    await cat.insert()
    return {"success": True, "data": _serialize(cat)}


# ── PUT /api/categories/{id} ──────────────────────────────────────────────────

@router.put("/{cat_id}")
async def update_category(cat_id: str, body: CategoryUpdate):
    cat = await _get_or_404(cat_id)
    if body.name  is not None: cat.name  = body.name.strip()
    if body.color is not None: cat.color = body.color
    if body.icon  is not None: cat.icon  = body.icon
    await cat.save()
    return {"success": True, "data": _serialize(cat)}


# ── DELETE /api/categories/{id} ───────────────────────────────────────────────

@router.delete("/{cat_id}")
async def delete_category(cat_id: str):
    cat = await _get_or_404(cat_id)
    await cat.delete()
    # Unlink all todos that belonged to this category
    await Todo.find(Todo.category.id == cat.id).update({"$set": {"category": None}})  # type: ignore
    return {"success": True, "message": "Category deleted and todos unlinked"}
